src/Aspire.Hosting.Azure.ServiceBus/AzureServiceBusExtensions.cs

```diff
@@ -124,19 +151,381 @@ public static IResourceBuilder<AzureServiceBusResource> AddQueue(this IResourceB
+    /// </summary>
+    /// <param name="builder">The Azure Service Bus resource builder.</param>
     /// <param name="topicName">The name of the topic.</param>
     /// <param name="subscriptionName">The name of the subscription.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
     public static IResourceBuilder<AzureServiceBusResource> AddSubscription(this IResourceBuilder<AzureServiceBusResource> builder, string topicName, string subscriptionName)
     {
-        builder.Resource.Subscriptions.Add((topicName, subscriptionName));
+        builder.WithTopic(topicName, topic =>
+        {
+            if (!topic.Subscriptions.Any(x => x.Name == subscriptionName))
+            {
+                topic.Subscriptions.Add(new ServiceBusSubscription(subscriptionName));
+            }
+        });
+
         return builder;
     }
+
+    /// <summary>
+    /// Configures an Azure Service Bus resource to be emulated. This resource requires an <see cref="AzureServiceBusResource"/> to be added to the application model.
+    /// </summary>
+    /// <remarks>
+    /// This version of the package defaults to the <inheritdoc cref="ServiceBusEmulatorContainerImageTags.Tag"/> tag of the <inheritdoc cref="ServiceBusEmulatorContainerImageTags.Registry"/>/<inheritdoc cref="ServiceBusEmulatorContainerImageTags.Image"/> container image.
+    /// </remarks>
+    /// <param name="builder">The Azure Service Bus resource builder.</param>
+    /// <param name="configureContainer">Callback that exposes underlying container used for emulation to allow for customization.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    /// <example>
+    /// The following example creates an Azure Service Bus resource that runs locally is an emulator and referencing that
+    /// resource in a .NET project.
+    /// <code lang="csharp">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    ///
+    /// var serviceBus = builder.AddAzureServiceBus("myservicebus")
+    ///    .RunAsEmulator()
+    ///    .AddQueue("queue");
+    ///
+    /// builder.AddProject&lt;Projects.InventoryService&gt;()
+    ///        .WithReference(serviceBus);
+    ///
+    /// builder.Build().Run();
+    /// </code>
+    /// </example>
+    public static IResourceBuilder<AzureServiceBusResource> RunAsEmulator(this IResourceBuilder<AzureServiceBusResource> builder, Action<IResourceBuilder<AzureServiceBusEmulatorResource>>? configureContainer = null)
+    {
+        if (builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            return builder;
+        }
+
+        // Add emulator container
+        var configHostFile = Path.Combine(Path.GetTempPath(), "AspireServiceBusEmulator", Path.GetRandomFileName() + ".json");
+
+        Directory.CreateDirectory(Path.GetDirectoryName(configHostFile)!);
+
+        var password = PasswordGenerator.Generate(16, true, true, true, true, 0, 0, 0, 0);
```

Comment:
Is generating a new password every time going to work well?

Similarly, do people expect to maintain data across runs? Other emulators have a WithDataVolume feature. (I'm not saying I think we need this, I'm just asking.)

Reply:
I didn't think about supporting this. Would be nice though. Can we punt the feature or do we assume we provide this anytime we can?

Reply:
I asked the team if that is a supported scenario (or if the db gets initialized on startup for instance).